### PR TITLE
fix(serverless): Patch to enable Secret filtering in KNative Serving

### PR DESCRIPTION
**Pull request by israel-hdez**
## Description
This prevents KNative's net-istio pods from being OOMKilled in environments with a large number of Secrets.

This would patch a Managed installation. Custom installations would be left untouched.

Fixes: https://issues.redhat.com/browse/RHOAIENG-1467

## How Has This Been Tested?
* Do a build of the operator and deploy it:
```sh
make image-build IMG=$YOUR_CONTAINER_REPO
make deploy IMG=$YOUR_CONTAINER_REPO
```

* Create the default DSCI
```sh
oc apply -f config/samples/dscinitialization_v1_dscinitialization.yaml
```

* Create a KServe DSC
```yaml
apiVersion: datasciencecluster.opendatahub.io/v1
kind: DataScienceCluster
metadata:
  name: kserve-dsc
spec:
  components:
    dashboard:
      managementState: Removed
    datasciencepipelines:
      managementState: Removed
    codeflare:
      managementState: Removed
    ray:
      managementState: Removed
    kserve:
      managementState: Managed
      serving:
        managementState: Managed
    modelmeshserving:
      managementState: Removed
    workbenches:
      managementState: Removed
```

* Check the net-istio-controller depoyment of KNative has the `ENABLE_SECRET_INFORMER_FILTERING_BY_CERT_UID` environment variable set to "true".
```sh
oc get deploy -n knative-serving net-istio-controller -o jsonpath='{.spec.template.spec.containers[*].env[*]}'
```

## Merge criteria:
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl
Comment (israel-hdez) [inline pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl:21]: @RetoCode @skonto Tagging you to confirm that this is what makes sense for the long-term.

**Comment by israel-hdez** [comment]
@bartoszmajsak Not adding you as a formal reviewer. Just wanted to ask this is the right way to do it so that also an upgrade would get the patched resource.

File: pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl
Comment (ReToCode) [reply pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl:21]: Yes it makes sense. We are working on making this the default anyway. Downstream it will be the default in the next release (1.32).

Please use `workloads` instead of `deployments`. `deployments` is deprecated and replaced with `workloads`.

Comment (israel-hdez) [reply pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl:21]: So, I changed, but there is a catch:

Being this a patch, I wonder what would happen for existing installs that need to be upgraded?
My guess is that since this is a patch, we would end up with a resource that would contain both `deployments` and `workloads` (assuming it would be accepted). So, not sure if it would be a good state....

**Comment by israel-hdez** [comment]
@VaishnaviHire @zdtsw Waiting for additional feedback, but this is becoming ready to go.

File: pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl
Comment (ReToCode) [reply pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl:21]: We do have that problem either way until we finally drop `deployments`. But `workloads` is applied later than `deployments`, so if we have conflicting values the `workloads` wins.  

Anything I forgot @skonto?

**Comment by zdtsw** [comment]
> @VaishnaviHire @zdtsw Waiting for additional feedback, but this is becoming ready to go.

in short, by add this env variable `ENABLE_SECRET_INFORMER_FILTERING_BY_CERT_UID=true`in net-istio controller 
 equals to adding annotation `serverless.openshift.io/enable-secret-informer-filtering: "true"`  in KnativeServing CR ?

File: pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl
Comment (skonto) [reply pkg/feature/templates/serverless/serving-net-istio-secret-filtering.patch.tmpl:21]: Deployments first and then workloads are merged in a slice and then they are applied in that order.

**Comment by ReToCode** [comment]
> in short, by add this env variable ENABLE_SECRET_INFORMER_FILTERING_BY_CERT_UID=truein net-istio controller
equals to adding annotation serverless.openshift.io/enable-secret-informer-filtering: "true" in KnativeServing CR ?

Yepp, the annotation will be removed in future releases.

**Comment by israel-hdez** [comment]
Alright, so this looks ready to go.

@zdtsw @VaishnaviHire when you are ready, this just awaits a review from you :slightly_smiling_face:

**Review by zdtsw — APPROVED** [review]

**Pull request merged**
